File: src/server/Model/Admin/Concept.cs
// Copyright (c) 2019, UW Medicine Research IT
// Developed by Nic Dobbins and Cliff Spital
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
using System;
using Model.Tagging;
using System.Collections.Generic;
using Model.Compiler;

namespace Model.Admin
{
    public class Concept
    {
        public Guid Id { get; set; }
        public ConceptUrn UniversalId { get; set; }
        public Guid? ParentId { get; set; }
        public Guid? RootId { get; set; }
        public string ExternalId { get; set; }
        public string ExternalParentId { get; set; }
        public int? SqlSetId { get; set; }
        public bool? IsNumeric { get; set; }
        public bool? IsParent { get; set; }
        public bool? IsPatientCountAutoCalculated { get; set; }
        public bool? IsSpecializable { get; set; }
        public string SqlSetWhere { get; set; }
        public string SqlFieldNumeric { get; set; }
        public string UiDisplayName { get; set; }
        public string UiDisplayText { get; set; }
        public string UiDisplaySubtext { get; set; }
        public string UiDisplayUnits { get; set; }
        public string UiDisplayTooltip { get; set; }
        public int? UiDisplayPatientCount { get; set; }
        public IEnumerable<ConceptPatientYearCount> UiDisplayPatientCountByYear { get; set; }
        public string UiNumericDefaultText { get; set; }

        public IEnumerable<ConceptConstraint> Constraints { get; set; }

        public IEnumerable<SpecializationGroupRelationship> SpecializationGroups { get; set; }
    }

    public class SpecializationGroupRelationship
    {
        public int SpecializationGroupId { get; set; }
        public int? OrderId { get; set; }
    }

    public enum ConstraintType
    {
        User = 1,
        Group = 2
    }

    public class ConceptConstraint
    {
        public Guid ConceptId { get; set; }
        public ConstraintType ConstraintId { get; set; }
        public string ConstraintValue { get; set; }

        public static ConstraintType TypeFrom(int code)
        {
            switch (code)
            {
                case 1:
                    return ConstraintType.User;
                case 2:
                    return ConstraintType.Group;
                default:
                    throw new InvalidOperationException($"{code} is not a valid ConstraintType.");
            }
        }
    }
}

File: src/ui-client/src/models/admin/Concept.ts
/* Copyright (c) 2022, UW Medicine Research IT, University of Washington
 * Developed by Nic Dobbins and Cliff Spital, CRIO Sean Mooney
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */ 

import { PatientCountPerYear } from "../concept/Concept";

export interface ConceptSqlSetDTO {
    id: number,
    isEncounterBased: boolean;
    isEventBased: boolean;
    sqlFieldDate?: string;
    sqlFieldEvent?: string;
    sqlSetFrom: string;
    eventId?: number;
}

export interface ConceptSqlSet extends ConceptSqlSetDTO {
    changed?: boolean;
    specializationGroups: Map<number,SpecializationGroup>;
    unsaved?: boolean;
}

export interface ConceptEvent {
    id: number;
    uiDisplayEventName: string;
    changed?: boolean;
    unsaved?: boolean;
}

export interface SpecializationGroupDTO {
    id: number;
    sqlSetId: number;
    specializations: Specialization[];
    uiDefaultText: string;
}

export interface SpecializationGroup {
    changed?: boolean;
    id: number;
    sqlSetId: number;
    specializations: Map<string,Specialization>;
    uiDefaultText: string;
    unsaved?: boolean;
}

export interface SpecializationGroupRelationship {
    specializationGroupId: number;
    orderId: number;
}

export interface SpecializationDTO {
    id: string;
    specializationGroupId: number;
    orderId?: number;
    sqlSetWhere: string;
    uiDisplayText: string;
    universalId?: string;
}

export interface Specialization extends SpecializationDTO {
    changed?: boolean;
    sqlSetId: number;
    unsaved?: boolean;
}

export interface Concept {
    id: string;
    parentId?: string;
    rootId: string;
    externalId?: string;
    externalParentId?: string;
    universalId?: string;
    isNumeric: boolean;
    isParent: boolean;
    isRoot: boolean;
    isPatientCountAutoCalculated: boolean;
    isSpecializable: boolean;
    isQueryable: boolean;
    constraints: Constraint[];
    specializationGroups: SpecializationGroupRelationship[];
    sqlSetId?: number;
    sqlSetWhere?: string;
    sqlFieldNumeric?: string;
    uiDisplayName: string;
    uiDisplayText: string;
    uiDisplaySubtext?: string;
    uiDisplayUnits?: string;
    uiDisplayTooltip?: string;
    uiDisplayPatientCount?: number;
    uiDisplayPatientCountByYear?: PatientCountPerYear[];
    uiDisplayRowOrder?: number;
    uiNumericDefaultText?: string;

    unsaved?: boolean;
}

export enum ConstraintType {
    User = 1,
    Group = 2
}

export interface Constraint {
    resourceId?: string;
    constraintId: ConstraintType;
    constraintValue: string;
}

export interface ConceptDeleteResponse {
    panelFilterCount: number;
    panelFilters: PanelFilterDependent[];
    queryCount: number;
    queries: QueryDependent[];
    conceptCount: number;
    concepts: ConceptDependent[];
}

export interface ConceptSqlSetDeleteResponse {
    conceptCount: number;
    concepts: ConceptDependent[];
    specializationGroupCount: number;
    specializationGroups: SpecializationGroupDependent[];
}

export interface SpecializationGroupDeleteResponse {
    conceptCount: number;
    concepts: ConceptDependent[];
}

interface PanelFilterDependent {
    id: number;
    uiDisplayText: string;
}

interface QueryDependent {
    id: string;
    universalId: string;
    name: string;
    owner: string;
}

interface ConceptDependent {
    id: string;
    universalId?: string;
    uiDisplayName: string;
}

interface SpecializationGroupDependent {
    id: string;
    uiDefaultText: string;
}